src/Pos/Txp/Toil/Utxo/Functions.hs

```diff
@@ -1,62 +1,232 @@
+    } deriving (Show)
+
+-- | CHECK: Verify Tx correctness using 'MonadUtxoRead'.
+-- Specifically there are the following checks:
+--
+-- * every input is a known unspent output;
+-- * sum of inputs >= sum of outputs;
+-- * every input has a proper witness verifying that input;
+-- * script witnesses have matching script versions;
+-- * if 'vtcVerifyVersions' is 'True', addresses' versions are verified
+--   to be known.
+--
+-- Note that 'verifyTx' doesn't attempt to verify scripts with versions
+-- higher than maximum script version we can handle. That's because we want
+-- blocks with such transactions to be accepted (to avoid hard
+-- forks). However, we won't include such transactions into blocks when we're
+-- creating a block.
 verifyTxUtxo
-    :: (MonadUtxoRead m, MonadError TxpVerFailure m)
-    => Bool
+    :: (MonadUtxoRead m, MonadError ToilVerFailure m)
+    => VTxContext
     -> TxAux
     -> m TxUndo
-verifyTxUtxo verifyVersions txaux = do
-    res <- verifyTx verifyVersions VTxGlobalContext utxoGet' txaux
-    case res of
-        Left errors -> throwError $ TxpInvalid $ T.intercalate ";\n" errors
-        Right undo  -> pure undo
+verifyTxUtxo ctx ta@(UnsafeTx {..}, witnesses, _) = do
+    verifyConsistency _txInputs witnesses
+    verResToMonadError (ToilInvalidOutputs . formatFirstError) $
+        verifyOutputs ctx ta
+    resolvedInputs <- mapM resolveInput _txInputs
+    verifySums resolvedInputs _txOutputs
+    verResToMonadError ToilInvalidInputs $
+        verifyInputs ctx resolvedInputs ta
+    return $ map snd resolvedInputs
+
+resolveInput
+    :: (MonadUtxoRead m, MonadError ToilVerFailure m)
+    => TxIn -> m (TxIn, TxOutAux)
+resolveInput txIn = (txIn, ) <$> (note (ToilNotUnspent txIn) =<< utxoGet txIn)
+
+verifySums
+    :: MonadError ToilVerFailure m
+    => NonEmpty (TxIn, TxOutAux) -> NonEmpty TxOut -> m ()
+verifySums resolvedInputs outputs =
+    when (outSum > inpSum) $
+    throwError $ ToilOutGTIn {tInputSum = inpSum, tOutputSum = outSum}
+  where
+    outSum = sumCoins $ map txOutValue outputs
+    inpSum = sumCoins $ map (txOutValue . toaOut . snd) resolvedInputs
+
+verifyConsistency :: MonadError ToilVerFailure m => NonEmpty TxIn -> TxWitness -> m ()
+verifyConsistency inputs witnesses
+    | length inputs == length witnesses = pass
+    | otherwise = throwError $ ToilInconsistentTxAux errMsg
   where
-    utxoGet' x = fmap VTxLocalContext <$> utxoGet x
+    errFmt = ("length of inputs != length of witnesses "%"("%int%" != "%int%")")
+    errMsg = sformat errFmt (length inputs) (length witnesses)
+
+verifyOutputs :: VTxContext -> TxAux -> VerificationRes
+verifyOutputs VTxContext {..} (UnsafeTx {..}, _, distrs)=
+    verifyGeneric $
+        [ ( length _txOutputs == length (getTxDistribution distrs)
+          , "length of outputs != length of tx distribution")
+        ]
+        ++
+        do (i, (TxOut{..}, d)) <-
```

Comment:
lul. don't you want to create function for this logic?
instead of `++ do` and half-triangle code